python/ray/node.py

```diff
@@ -559,20 +560,19 @@ def create_redis_client(self):
 
     def get_gcs_client(self):
         if self._gcs_client is None:
-            num_retries = NUM_REDIS_GET_RETRIES
-            for i in range(num_retries):
+            for _ in range(NUM_REDIS_GET_RETRIES):
+                gcs_address = None
+                last_ex = None
                 try:
-                    if use_gcs_for_bootstrap():
-                        self._gcs_client = GcsClient(address=self.gcs_address)
-                    else:
-                        self._gcs_client = GcsClient.create_from_redis(
-                            self.create_redis_client())
```

Comment:
We should keep this if-else since we are not removing the old logic.

pass redis cli into gcs client could help gcs to obtain the new addr.

Reply:
`self.gcs_address` also uses Redis to fetch GCS address, when not using GCS bootstrapping. I think reducing branching simplifies the logic.

Reply:
it's not about self.gcs_address.

https://github.com/ray-project/ray/blob/master/python/ray/_private/gcs_utils.py#L172-L174

Here is the place where reconnect happens. So when gcs is down, and up again, the address will change. so we need to always fetch gcs address for redis mode.

Reply:
Not sure I understand. GcsClient creates a GcsChannel even when initialized with address, and the reconnect logic above still runs inside GcsChannel, right?

Reply:
Discussed off thread. I double checked that GCS starts at a cached port regardless of Redis or GCS bootstrapping mode. Added a comment to GcsChannel to mention this fact.

Reply:
Btw code to use cached port: https://github.com/ray-project/ray/blob/e9068c45faf2928cd8e38f6d38c5260cb06130d2/python/ray/node.py#L218-L226